### PR TITLE
MacOS file type filter in native file dialog

### LINKED ISSUE #5216: There makes no sense using defaultExtension of SaveFileDialog
just like what in the following image:

image

it is no extension info when call saveFileDialog entity's showDialog function, it return the file name with no extensions.

The develop environment is macOS, Rider IDE.

### PR DESCRIPTION
**Pull request by maxkatz6**
## What does the pull request do?

1. Adds file type chooser to the native file dialog
2. Fixes support for file extensions
3. Ensures file formats priority to be "File extension > Apple Type UID > MimeType".
4. Default file extension is also respected as default selected file format + macOS automatically appends this extension to the file name.

## What is the current behavior?

It's not possible for the user to select file type filter.
File extension file filters don't work.

## How was the solution implemented (if it's not obvious)?

https://github.com/chromium/chromium/blob/49f63f17ac50384e9f367622dd5baf9701d4da3f/components/remote_cocoa/app_shim/select_file_dialog_bridge.mm#L65

## Fixed issues

Fixes #5216

### CODE REVIEW

**Comment by maxkatz6** [comment]
https://github.com/AvaloniaUI/Avalonia/assets/3163374/fb8121d1-c2f2-4046-af9e-535889135dfa

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0039550-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

File: src/Avalonia.Native/AvnString.cs [hunk 1]
```diff
@@ -61,6 +63,21 @@ private unsafe void EnsureNative()
...
+
+        public AvnStringArray(IReadOnlyList<string> items)
+        {
+            _items = items;
+        }
+
+        public string[] ToStringArray() => (_items as string[] ?? _items.ToArray());
+
+        public uint Count => (uint)_items.Count;
+        public IAvnString Get(uint index) => _items[(int)index].ToAvnString();
```
Comment (kekekeks) [inline src/Avalonia.Native/AvnString.cs:79]: Please check if AvnString instance is actually getting disposed.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0040072-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

File: src/Avalonia.Native/AvnString.cs (hunk 1 above)
Comment (maxkatz6) [reply src/Avalonia.Native/AvnString.cs]: Fixed with GetNSArrayOfStringsAndRelease + explicit disposal of strings

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0040084-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

**Review by jmacato — APPROVED** [review]
Confirmed crash is fixed, tested locally LGTM

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0040146-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

**Pull request merged**